src/database/database.cpp

```diff
@@ -0,0 +1,1099 @@
+        *mSession << "DROP TABLE IF EXISTS layers;", now;
+        *mSession << "DROP TABLE IF EXISTS instances;", now;
+
+    } catch (const Poco::Exception& e) {
+        return Error(AOS_ERROR_WRAP(ErrorEnum::eFailed), e.what());
+    }
+
+    return ErrorEnum::eNone;
+}
+
+Error Database::CreateConfigTable()
+{
+    auto [tableExists, err] = TableExist("config");
+
+    if (!err.IsNone()) {
+        return err;
+    }
+
+    if (tableExists) {
+        uint64_t operationVersion {0};
+
+        Tie(operationVersion, err) = GetOperationVersion();
+        if (!err.IsNone()) {
+            return err;
+        }
+
+        if (operationVersion == sm::launcher::Launcher::cOperationVersion) {
+            return ErrorEnum::eNone;
+        }
+
+        if (err = DropAllTables(); !err.IsNone()) {
+            return Error(AOS_ERROR_WRAP(err), "failed to drop all tables");
+        }
+    }
+
+    *mSession << "CREATE TABLE config ("
+                 "operationVersion INTEGER, "
+                 "cursor TEXT, "
+                 "envvars TEXT, "
+                 "onlineTime TIMESTAMP);",
+        now;
+
+    try {
+        *mSession << "INSERT INTO config ("
+                     "operationVersion, "
+                     "onlineTime) values(?, ?);",
+            bind(Poco::Tuple<uint32_t, uint64_t>(sm::launcher::Launcher::cOperationVersion, Time::Now().UnixNano())),
+            now;
+    } catch (const Poco::Exception& e) {
+        return AOS_ERROR_WRAP(ErrorEnum::eFailed);
+    }
+
+    return ErrorEnum::eNone;
+}
+
+void Database::CreateTables()
+{
+    *mSession << "CREATE TABLE IF NOT EXISTS network ("
+                 "networkID TEXT NOT NULL PRIMARY KEY, "
+                 "ip TEXT, "
+                 "subnet TEXT, "
+                 "vlanID INTEGER, "
+                 "vlanIfName TEXT);",
+        now;
+
+    *mSession << "CREATE TABLE IF NOT EXISTS services ("
+                 "id TEXT NOT NULL , "
+                 "version TEXT, "
+                 "providerID TEXT, "
+                 "imagePath TEXT, "
+                 "manifestDigest BLOB, "
+                 "cached INTEGER, "
+                 "timestamp TIMESTAMP, "
+                 "size INTEGER, "
+                 "GID INTEGER, "
+                 "PRIMARY KEY(id, version));",
+        now;
+
+    *mSession << "CREATE TABLE IF NOT EXISTS trafficmonitor ("
+                 "chain TEXT NOT NULL PRIMARY KEY, "
+                 "time TIMESTAMP, "
+                 "value INTEGER)",
+        now;
+
+    *mSession << "CREATE TABLE IF NOT EXISTS layers ("
+                 "digest TEXT NOT NULL PRIMARY KEY, "
+                 "layerId TEXT, "
+                 "path TEXT, "
+                 "osVersion TEXT, "
+                 "version TEXT, "
+                 "timestamp TIMESTAMP, "
+                 "cached INTEGER, "
+                 "size INTEGER)",
+        now;
+
+    *mSession << "CREATE TABLE IF NOT EXISTS instances ("
+                 "serviceID TEXT, "
```

Comment:
serviceID, subjectID, instance are key fields

Reply:
It's declared "PRIMARY KEY(serviceID, subjectID, instance))"